### ISSUE TITLE
Create peephole-like optimizer stage that applies Simplification rules

### ISSUE THREAD

**Issue by chriseth** [ethereum/solidity#4674]
The `RuleList.h` contains a list of simplification rules that can be applied to sequences (or rather expressions) of opcodes. Examples are simple arithmetics on constants or other simplifications that can be applied regardless of the context:

 - `push1 7 push1 8 add` is reduced to `push1 15`
 - `dup1 dup1 xor` which is reduced to `push1 0`

These rules are currently used in two places: The first one is the new optimizer that will be used on Solidity inline assembly, and the second one is the so-called "common subexpression eliminator" that uses these rules both to perform equality checks during symbolic execution and to actually perform the simplifications.

Since the common subexpression eliminator is a rather complex module and sometimes fails to run on certain code snippets, it makes sense to also run the simplification rules in the old optimizer just on sequences of opcodes prior to running the full subexpression eliminator.

Since the rule list can be proven correct, it might also make sense to run this module even if the optimizer is turned off (just like the current peephole optimizer).

This optimizer stage should be similar to the peephole optimizer: It takes a vector of assembly items and outputs a vector of assembly items.

There are two files called `SimplificationRules.h/cpp`, one for the yul optimizer and the other for the common subexpression eliminator. A third one should probably be created for this new module.

The concept of "match groups" should probably be ignored for the first iteration of solving this issue. They have two purposes: First, they are used to check that two "items" are identical (e.g. for the rule `xor(x, x) -> 0`) and secondly, they are used in the execution of the replacement pattern (e.g. for the rule `or(x, x) -> x`). This module should only do local replacements in the stream of opcodes. This means the xor-pattern above could result in `dup1 xor` -> `pop push1 0`, but (at least its initial version) should not perform replacements if it can be inferred that the two top-most stack elements are identical, but you would need non-local information to do that. Also note that the replacement above specifically uses `pop` instead of omitting the preceding opcode, because that opcode might have side-effects. There are other optimizer stages (even inside the peephole optimizer) that will remove the opcode preceding a pop if it does not have side-effects.

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__This issue now has a funding of 525.0 DAI (525.0 USD @ $1.0/DAI)  attached to it.__

 * If you would like to work on this issue you can 'start work' [on the Gitcoin Issue Details page](https://gitcoin.co/issue/ethereum/solidity/4674/907).
* Want to chip in? Add your own contribution [here](https://gitcoin.co/issue/ethereum/solidity/4674/907).
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
* $24,669.07 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by chriseth** [comment]
@meowingtwurtle is taking a shot at this.

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__Workers have applied to start work__.


These users each claimed they can complete the work by 12 months from now.
Please review their action plans below:


1. [meowingtwurtle](https://gitcoin.co/profile/meowingtwurtle) has applied to start work _(Funders only: [approve worker](https://gitcoin.co/issue/ethereum/solidity/4674/907?mutate_worker_action=approve&worker=meowingtwurtle) | [reject worker](https://gitcoin.co/issue/ethereum/solidity/4674/907?mutate_worker_action=reject&worker=meowingtwurtle))_. 
    
    agreed to by @chriseth

This is filler because gitcoin requires it.
This is filler because gitcoin requires it.

Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/ethereum/solidity/4674/907).

**Comment by gitcoinbot** [comment]
@meowingtwurtle Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] warning (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/ethereum/solidity/4674/907?snooze=1) | [3 days](https://gitcoin.co/issue/ethereum/solidity/4674/907?snooze=3) | [5 days](https://gitcoin.co/issue/ethereum/solidity/4674/907?snooze=5) | [10 days](https://gitcoin.co/issue/ethereum/solidity/4674/907?snooze=10) | [100 days](https://gitcoin.co/issue/ethereum/solidity/4674/907?snooze=100)

**Comment by gitcoinbot** [comment]
@meowingtwurtle Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] warning (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/ethereum/solidity/4674/907?snooze=1) | [3 days](https://gitcoin.co/issue/ethereum/solidity/4674/907?snooze=3) | [5 days](https://gitcoin.co/issue/ethereum/solidity/4674/907?snooze=5) | [10 days](https://gitcoin.co/issue/ethereum/solidity/4674/907?snooze=10) | [100 days](https://gitcoin.co/issue/ethereum/solidity/4674/907?snooze=100)

**Comment by vs77bb** [comment]
Hi @meowingtwurtle are you still working here? I saw a lot of commentary on the PR last week, so I'll snooze Gitcoin Bot reminders. Please do let us know if you're not planning on taking this forward for any reason 🙂

**Comment by randomnetcat** [comment]
Sorry, I've just gone back to school and have been a bit busy. chriseth has also been unavailable, and he was the one giving me most of the feedback. I do still plan to complete this.

**Comment by gitcoinbot** [comment]
@meowingtwurtle Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] warning (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/ethereum/solidity/4674/907?snooze=1) | [3 days](https://gitcoin.co/issue/ethereum/solidity/4674/907?snooze=3) | [5 days](https://gitcoin.co/issue/ethereum/solidity/4674/907?snooze=5) | [10 days](https://gitcoin.co/issue/ethereum/solidity/4674/907?snooze=10) | [100 days](https://gitcoin.co/issue/ethereum/solidity/4674/907?snooze=100)

**Comment by gitcoinbot** [comment]
@meowingtwurtle Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] warning (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/ethereum/solidity/4674/907?snooze=1) | [3 days](https://gitcoin.co/issue/ethereum/solidity/4674/907?snooze=3) | [5 days](https://gitcoin.co/issue/ethereum/solidity/4674/907?snooze=5) | [10 days](https://gitcoin.co/issue/ethereum/solidity/4674/907?snooze=10) | [100 days](https://gitcoin.co/issue/ethereum/solidity/4674/907?snooze=100)

**Comment by randomnetcat** [comment]
Yes, I am.

**Comment by meowingtwurtle** [comment]
@chriseth (sorry, thought I sent this before). The functions on the top are useful in more than just the match function. They also allow us to decompose a larger opcode stream into its expressions and recursively apply rules.

**Comment by liangdzou** [comment]
Actually, there are some papers on peephole optimization for EVM instructions. You may want to refer to "Towards Saving Money in Using Smart Contracts" in ICSE2018 or "Under-Optimized Smart Contracts
Devour Your Money" in SANER2017.

These two papers are done by the same group, where the ICSE2018 version is the newest result but without less explanation on why. The basic idea is to apply rules to reduce EVM instructions. However, since it is work on EVM instructions, they have to fix the jump location problem. I think this will be much easier if we implement on Solidity's representation since labels are used here.

**Comment by liangdzou** [comment]
@meowingtwurtle @chriseth 
If you guys are too busy, I would like to have a try.
My employee actually supports me to do some blockchain related open source projects (at least for a few months).

**Comment by randomnetcat** [comment]
There is already existing code here. I am waiting for more review from @chriseth here. I agree with him that the code needs to be simplified, but I really am not sure specifically how he wants that to be done.

**Comment by liangdzou** [comment]
I would regard this as a long-term project and have a more detailed plan before implementation. For instance, I think it would be better to read those two papers first and investigate if the patterns given in the paper are useful. Then, to discuss if the pattern solve the most of the problems that would be solved by peephole optimization. If not, we should try to extend the list. Then, to discuss the order to apply the patterns and discuss if we should apply the pattern list recursively. Finally, the implementation.

Also, I think it would be better to have a set of high quality performance tests before doing anything else.

**Comment by randomnetcat** [comment]
The specification was for this to be a relatively simple module that applies the pre-existing SimplificationRules from CSE. It is a nongoal for this to be a complete optimizer for all of EVMASM.

**Comment by chriseth** [comment]
@liangdzou do you happen to have a publicly available version of that paper? I heard only of 5 rules and we already have all of them. Please take a look at https://github.com/ethereum/solidity/blob/develop/libevmasm/RuleList.h to find already existing rules. This issue is not about finding new rules but rather employing them in a third part of the optimizer.

**Comment by liangdzou** [comment]
This is the link given by the author. (https://docs.google.com/document/d/1VvKI7BEwq2EFcpwzqEe1-vT7a042cwAnYzbvT7s2Ijw/edit)
The following is the content.

The operation sequence before → denotes an anti-pattern and the sequence after → is the code after optimization (delete indicates the anti-pattern can be removed). 

P1 — {swap(X), swap(X)} → delete, 1 ≤ X ≤ 16
() indicates a number inside instead of a part of an operation. swap(X) exchanges the 1st and the (𝑋 + 1)th stack items. The two consecutive swap(X)s can be removed since the execution of them has no effect. The anti-pattern accounts for two bytes (e.g., the bytecode of swap1 is 0x90) and its deployment wastes 136 units of gas (every non-zero bytecode requires 68 units). The execution of this pattern wastes 6 units of gas because the execution of a swap(X) needs 3 units. 

P2 — {M consecutive jumpdests} → {jumpdest}, M ≥ 2
jumpdest marks a valid jump target (i.e., any valid jump operation must jump to a jumpdest). M consecutive jumpdests can be replaced with one because they point to the same operation (i.e., the operation after the final jumpdest). The deployment and invocation of this anti-pattern waste 68 × (𝑀 − 1) units and (𝑀 − 1) units of gas (the execution of a jumpdest needs 1 unit of gas), respectively.

P3 — {OP, pop} → {pop}, OP ∈ {iszero, not, balance, calldataload, extcodesize, blockhash, mload, sload}
OP consumes the top stack item, produces an outcome, and pushes the outcome on stack. pop removes the top stack item. This anti-pattern is semantically equivalent with one pop because the outcome of OP is immediately removed by pop. Due to page limit, we cannot explain the semantics of every EVM operation, and interested readers can refer to Ethereum’s yellow paper. The anti-pattern wastes 68 units of gas in deployment, and the amount of gas wasted by executing the anti-pattern instance depends on gas_cost of OP.  

P4 — {OP, pop} → {pop, pop}, OP ∈ {add, sub, mul, div, sdiv, mod, smod, exp, sigextnd, lt, gt, slt, sgt, eq, and, or, xor, byte, sha3}
This anti-pattern is similar with anti-pattern 3, but this OP consumes the top two stack items, produces an outcome, and pushes the outcome on stack. This anti-pattern is semantically equivalent with two consecutive pops because the the top 2 stack items are removed by two pops. The anti-pattern wastes 0 units of gas in deployment because both the anti-pattern and the efficient code consist of two non-zero bytes. The amount of gas wasted by executing the anti-pattern instance depends on gas_cost of OP.

P5 — {OP, pop} → {pop, pop, pop}, OP ∈ {addmod, mulmod}
This anti-pattern is similar with anti-pattern 3, but this OP consumes the top three stack items, produces an outcome, and pushes the outcome on stack. This anti-pattern is semantically equivalent with three consecutive pops because the top 3 stack items are removed by pops. The anti-pattern is the only one which increases 68 units of gas in deployment after replacing with the efficient code. However, the three consecutive pops is more efficient than pop, pop is executing since a pop needs 2 units of gas and either addmod or mulmod needs 8 units of gas. Hence, the execution of this anti-pattern wastes 4 units of gas. Since a smart contract can be executed many times but can be deployed only once, OP, pop is considered to be less efficient than three pops.

P6 — {OP, pop} →  delete, OP ∈ {address, origin, caller, callvalue, calldatasize, codesize, gasprice, coinbase, timestamp, number, difficulty, gaslimit, pc, msize, gas}
This anti-pattern is similar with anti-pattern 3, but this OP doesn’t consume the stack item, but produces an outcome, and pushes the outcome on stack. Therefore, this anti-pattern can be deleted because the outcome of OP is immediately removed by pop. The anti-pattern wastes 136 units of gas in deployment, and the amount of gas wasted by executing the anti-pattern instance depends on gas_cost of OP.

P7 — {swap1, swap(X), OP, dup(X), OP} → {dup2, swap(X+1), OP, OP}, 2 ≤ X ≤ 15, OP ∈ {add, mul, and, or, xor} 
dup(X) duplicates Xth stack item on stack top. Let the first three stack items be i0, i1 and i2, X be 2, and OP be add. After executing this anti-pattern, the number of stack items decreases by 1 and the top two items become i0 + i1 + i2, i1. Hence, one swap can be removed since OP (e.g. add) is commutative. The anti-pattern wastes 68 units and 3 units of gas for deployment and execution of a swap, respectively.

P8 — {OP, stop} → {stop}, OP can be any operation except jumpdest, jump, jumpi and all operations that change storage.
stop halts the execution of the smart contract and then stack and memory are cleared. For example, let OP be add, after the execution of stop, the outcome of add on stack will be discarded, and hence add can be removed. On the other hand, those operations involved in control flow transfer or changing the permanent storage space cannot be removed.

P9 — {swap(X), dup(X+1), swap1} → {dup1, swap(X+1)}, 1 ≤ X ≤ 15
Let the top two stack items be i0 and i1, X be 1, after the execution of this anti-pattern, the number of stack items increases by 1 and the top three stack items become i1, i0, and i0. Therefore, one swap operation can be removed. The deployment and invocation of this anti-pattern wastes 68 units and 3 units of gas, respectively. 

P10 — {push(X), swap(Y), Y consecutive pops} → {Y consecutive pops, push(X)}, 1 ≤ X ≤ 32, 1 ≤ Y ≤ 16
As an example, let the top stack item be i0 and the value pushed by the first operation, push1 is i1, and Y be 1, after the execution of this anti-pattern, the top stack item becomes i1. Hence, this anti-pattern can be replaced with pop, push1. Therefore, one swap operation can be removed. The deployment and invocation of this anti-pattern wastes 68 units and 3 units of gas, respectively.

P11 — {swap(X), X+1 consecutive pops} → {X+1 consecutive pops}, 1 ≤ X ≤ 16
Let X be 1, since two consecutive pops pop top two stack items, the swap1 has no effect. The deployment and invocation of this anti-pattern wastes 68 units and 3 units of gas, respectively.

P12 — {swap(X), swap(Y), Y consecutive pops} →  {X consecutive pops, swap(Y-X), (Y-X) consecutive pops}, 1 ≤ X ≤ 15, X < Y
Let the top three stack items be i0, i1, and i2, X be 1 and Y be 2, after the execution of this anti-pattern, the number of stack items reduces by two and the top stack item becomes i1. One swap operation can be removed because the two consecutive pops cancel the effect of that swap. The deployment and invocation of this anti-pattern wastes 68 units and 3 units of gas, respectively. 

P13 — {X consecutive push(N)s, Y consecutive pops} → {(X-Y) consecutive push(N)s, if X > Y; (Y-X) consecutive pops, otherwise}, 1 ≤ N ≤ 32
A pop immediately after a push can be removed because they have no effect. To investigate gas waste, we consider the following two situations. First, push(N) outnumbers pop. In this case, all Y pop operations and Y push(N) operations can be removed. Therefore, the deployment and invocation of the anti-pattern waste 136Y (68×2Y) units and 5Y (3Y+2Y) units of gas. Otherwise, all X push(N) operations and X pop operations can be removed. In this case, the deployment and invocation of the anti-pattern waste 136X (68×2X) units and 5X (3X+2X) units of gas.

P14 — {X consecutive dup(N)s, Y consecutive pops} → {(X-Y) consecutive dup(N)s, if X > Y; (Y-X) consecutive pops, otherwise}, 1 ≤ N ≤ 16
This anti-pattern is similar with anti-pattern 13, because dup(N) duplicates the Nth stack item on stack, whose effect can be cancelled by pop. Similarly, the anti-pattern wastes 136Y units and 5Y units of gas in deployment and invocation, if X > Y; and 136X units and 5X units of gas, if X ≤ Y.

P15 — {dup(X), swap(X)} → {dup(X)}, 1 ≤ X ≤ 16
Let X be 1, dup1 duplicates the 1st stack item, so the following swap1 has no effect. The deployment and invocation of this anti-pattern wastes 68 units and 3 units of gas, respectively.

P16 — {swap1, swap2, …, swap(X), swap(X-1), …, swap1, X-1 consecutive pops} → {X-1 consecutive pops, swap1}, 2 ≤ X ≤ 16
Let the top three stack items be i0, i1 and i2, and X be 2, after the execution of this anti-pattern, the number of stack items reduces by one and the top two stack items become i2, i1. By executing pop before the sequence of swap operations, fewer swap operations are needed because swap(X) exchanges the top item with the (X+1)th stack item. In other words, EVM cannot adjust the order of non-top stack items directly; instead, it has to move one item on stack top and then exchange it with another stack item by executing several swaps. The optimized operation sequence removes 2X-2 redundant swap operations. Hence, the anti-pattern wastes 136(X-1) units and 6(X-1) units of gas in deployment and invocation, respectively.

P17 — {swap1, OP} → {OP}, OP ∈ {add, mul, and, or, xor}
Since OP is commutative, swap1 has no effect. Therefore, the deployment of the anti-pattern wastes 68 units and the amount of wasted gas in invocation is 3 units.

P18 — {OP, iszero, iszero} → {OP}, OP ∈ {lt, gt, slt, sgt, eq}
As an example, the outcome of gt is either 1 or 0, depending on whether the 1st stack item is larger than the 2nd stack item. iszero sets the 1st stack item to 1 if it is 0, or 0 otherwise. Therefore, two consecutive iszeros do not change the outcome of OP. Therefore, the deployment and invocation of the anti-pattern wastes 136 units and 6 units of gas, respectively.

P19 — {N consecutive push(X)，M consecutive swap(Y)} → {N consecutive push(X)}, Y < N, 1 ≤ X ≤ 32, 1 ≤ Y ≤ 16
push(X) are the only operations that has operands, let N be 2 and Y be 1, so the operation sequence push(X), push(X), swap1 can be optimized as push(X), push(X). That is to say, by adjusting the order of the operands of push operations, the swap operation can be removed. Therefore, the anti-pattern wastes 68M units and 3M units of gas in deployment and invocation, respectively.

P20 — {push(X), swap(Y), push(M), swap1} → {push(M), push(X), swap(Y+1)}, 1 ≤ X ≤ 32, 1 ≤ Y ≤ 15, 1 ≤ M ≤ 32
For the same reason with pattern 19, one swap operation of this anti-pattern can be removed, and hence it wastes 68 units and 3 units of gas in deployment and invocation, respectively.

P21 — {consecutive X push(N), dup(Y), swap(Z)} → {combination of X push(N) and dup(M)}, Y<=X, Z<=X, M<=X, 1 ≤ N ≤ 32, 1 ≤ Y ≤ 16, 1 ≤ Z ≤ 16, 1 ≤ M ≤ 16
For the same reason with pattern 19, the swap operation of this anti-pattern can be removed, and hence the deployment and invocation of this anti-pattern wastes 68 units and 3 units of gas, respectively. 

P22 — {swap(N), M consecutive OPs，(N+M) consecutive and same OP'} → {M consecutive OPs，(N+M) consecutive and same OP'}, OP ∈ {push(X), dup(Y)}, OP' ∈ {add, mul, and, or, xor}, 1 ≤ X ≤ 32, 1 ≤ N ≤ 16, 1 ≤ Y ≤ 16
Since OP' is commutative and both push(X) and dup(Y) increase the number of stack items by one, swap(N) in this anti-pattern can be removed. Therefore, the deployment and invocation of this anti-pattern wastes 68 units and 3 units of gas, respectively.

P23 — {dup1, swap(X), dup2, swap1} → {dup1, dup2, swap(X+1)}, 1 ≤ X ≤ 15
Let the first two stack items be i0, i1, and X be 2, after the execution of this anti-pattern, the number of stack items increases by two and the top four stack items become i1, i0, i0, i0. The optimized operation sequence has the same effect and has one fewer operations. Therefore, the deployment and invocation of this anti-pattern wastes 68 units and 3 units of gas, respectively.

P24 — {dup(X), swap(X-1), swap1, dup(X), swap1} → {dup(X), dup1, swap(X), swap2}, 3 ≤ X ≤ 16
Let the first three stack items be i0, i1, i2, and X be 3, after the execution of this anti-pattern, the number of stack items increases by two and the top five stack items become i0, i2, i1, i2, i2. One swap operation can be removed, and hence the deployment and invocation of this anti-pattern wastes 68 units and 3 units of gas, respectively.

**Comment by liangdzou** [comment]
I am interested in working on implementing those rules. Maybe started with building a large set of performance tests. Shall I create an issue for it?

**Comment by liangdzou** [comment]
@chriseth Do you think implementing the rules may be a better start.

**Comment by gitcoinbot** [comment]
@meowingtwurtle Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] warning (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/ethereum/solidity/4674/907?snooze=1) | [3 days](https://gitcoin.co/issue/ethereum/solidity/4674/907?snooze=3) | [5 days](https://gitcoin.co/issue/ethereum/solidity/4674/907?snooze=5) | [10 days](https://gitcoin.co/issue/ethereum/solidity/4674/907?snooze=10) | [100 days](https://gitcoin.co/issue/ethereum/solidity/4674/907?snooze=100)

**Comment by gitcoinbot** [comment]
@meowingtwurtle Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] warning (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/ethereum/solidity/4674/907?snooze=1) | [3 days](https://gitcoin.co/issue/ethereum/solidity/4674/907?snooze=3) | [5 days](https://gitcoin.co/issue/ethereum/solidity/4674/907?snooze=5) | [10 days](https://gitcoin.co/issue/ethereum/solidity/4674/907?snooze=10) | [100 days](https://gitcoin.co/issue/ethereum/solidity/4674/907?snooze=100)

**Comment by liangdzou** [comment]
@meowingtwurtle Can you submit your PR, we can work together to get it ready. :-)

**Comment by randomnetcat** [comment]
#4795. Again, it is waiting for more code review, as the core team is preoccupied with the next release.

**Comment by liangdzou** [comment]
thanks :-) I will take a look at this weekend.

**Comment by chriseth** [comment]
My general comment about #4795 still stands: It is much too complicated, at least I currently do not understand the helper functions.

**Comment by gitcoinbot** [comment]
@meowingtwurtle Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] warning (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/ethereum/solidity/4674/907?snooze=1) | [3 days](https://gitcoin.co/issue/ethereum/solidity/4674/907?snooze=3) | [5 days](https://gitcoin.co/issue/ethereum/solidity/4674/907?snooze=5) | [10 days](https://gitcoin.co/issue/ethereum/solidity/4674/907?snooze=10) | [100 days](https://gitcoin.co/issue/ethereum/solidity/4674/907?snooze=100)

**Comment by randomnetcat** [comment]
Aye.

**Comment by liangdzou** [comment]
I think it is reasonable to have a document on the PR to show some general thoughts about the implementation. For instance, the document needs to show what is the problem and how the PR solves the problem.

Meanwhile, a PR usually comes with some test cases, especially for new features. The test cases can show the cases that have been considered. It would be better to add the test cases to the test suite of Solidity.

I have some issues to deal with recently, not sure how much time I can spend here. :-)

**Comment by vs77bb** [comment]
@chriseth Would it make sense to return this to the crowd for another approach? @meowingtwurtle, we're happy to pay you a part of the bounty for your time here.

**Comment by randomnetcat** [comment]
Sorry, meant to reply much earlier. If you guys want to have someone else look into this, by all means go ahead. I'm happy to give anything I've learned.

**Comment by ChrisCates** [comment]
No worries @meowingtwurtle, we can pick up where you left off.
Seems like a lot of math work is involved here.

@liangdzou, are you still interested in creating a pull request?

**Comment by liangdzou** [comment]
@ChrisCates You can feel free to work on this topic. I won't have time for this at the moment.

**Comment by ChrisCates** [comment]
@liangdzou, no worries. I'll see what I can pull up soon.

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. Started 3. Submitted 4. **Done** 

<hr>

__Work has been started__.


These users each claimed they can complete the work by 2 months, 2 weeks from now.
Please review their action plans below:


**1) [mihairaulea](https://gitcoin.co/profile/mihairaulea) has been approved to start work.**

This sounds very interesting, read through the issue :) Also read the papers "Towards Saving Money in Using Smart Contracts" in ICSE2018 or "Under-Optimized Smart Contracts Devour Your Money" in SANER2017. I understand the rules are already written, and this is about using them in the 3rd stage of the optimizer. I think the reward is a bit low, but i trust that stellar work will be rewarded properly :D

Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/ethereum/solidity/4674/907).

**Comment by ChrisCates** [comment]
@mihairaulea, let me know if you want to work on it together. Ultimately I think this is more of a long-term project where we would want to do some deep auditing and the initial bounty is just to get the ball rolling and discussion flowing.

**Comment by chriseth** [comment]
Note that this should be a very very simple component for the start. The existing pull request is already way too complicated. I would be more than happy about a peephole optimizer that applies the existing ruleset to change `iszero iszero iszero` to `iszero`.

**Comment by mihairaulea** [comment]
@chriseth this is a bit confusing. just saw this. can i work on it? i have updated my notification settings, so i will not miss future conversations.

**Comment by chriseth** [comment]
Sure, please go ahead!

**Comment by kuhnchris** [comment]
Hey there @mihairaulea and @chriseth any progress on this issue in the last month? Need any support from our side or are you good to go, I see the thread is pretty long at this point. 
Feel free to contact us if you have any issue from the Gitcoin side of things!

Thanks,
Chris

**Comment by chriseth** [comment]
@vs77bb could you please award half of the bounty to @meowingtwurtle for the initial work he has done on this? Thanks!

**Comment by randomnetcat** [comment]
[Note, I changed my name: I was formerly @meowingtwurtle. As proof, that account has in its bio a message saying that I am its current user. I am happy to provide other proof if need be]

I hate to ask for this, @vs77bb, but I really could use this money (having hardware issues with my laptop; always fun).

Ethereum address: `0x6a58780f223D041cD3fDb6837b46bDb0d7335760`

**Comment by kuhnchris** [comment]
I'll see if I can get the message over to the responsible people, I'll keep you posted!

**Comment by randomnetcat** [comment]
Thanks.

**Comment by gitcoinbot** [comment]
⚡️ A tip worth 262.50000  DAI (262.5 USD @ $1.0/DAI) has been granted to @meowingtwurtle for this issue from @vs77bb. ⚡️ 

Nice work @meowingtwurtle! Your tip has automatically been deposited in the ETH address we have on file.

 * $86639.71 in Funded OSS Work Available at: https://gitcoin.co/explorer
 * Incentivize contributions to your repo: <a href='https://gitcoin.co/tip'>Send a Tip</a> or <a href='https://gitcoin.co/funding/new'>Fund a PR</a>
 * No Email? Get help on the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>

**Comment by vs77bb** [comment]
@random-internet-cat it should be paid out to your ETH Address. Let us know if you have any issues!

**Comment by randomnetcat** [comment]
Thank you!

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. Started 3. Submitted 4. **Done** 

<hr>

__This Bounty has been completed.__

Additional Tips for this Bounty:
* vs77bb tipped 262.5000 DAI worth 262.5 USD to meowingtwurtle.

<hr>

* Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/ethereum/solidity/4674/907)
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
* $86,040.18 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)
